posts/2024-08-26-webassembly-targets-change-in-default-target-features.md

```diff
@@ -0,0 +1,252 @@
+through [`wasm-opt`] these 5-byte encodings of index 0 are all shrunk to a
+single byte.
+
+## Enabling Multi-Value by Default
+
+The second feature enabled by default in LLVM 19 is `multivalue`. The
+[multi-value proposal to WebAssembly][multi-value] enables functions to have
+more than one return value for example. WebAssembly instructions are
+additionally allowed to have more than one return value as well. This proposal
+is one of the first to get merged into the WebAssembly specification after the
+original MVP and has been implemented in many engines for quite some time.
+
+The consequences of enabling this feature by default in LLVM are more minor for
+Rust, however, than enabling the `reference-types` feature by default. LLVM's
+default C ABI for WebAssembly code is not changing even when `multivalue` is
+enabled.  Additionally Rust's `extern "C"` ABI is not changing either and
+continues to match LLVM's. Despite this though the change has the possibility
+of still affecting Nightly users of Rust.
+
+Rust for some time has supported an `extern "wasm"` ABI on Nightly which was an
+experimental means of exposing the ability of defining a function in Rust which
+returned multiple values (e.g. used the multi-value proposal). Due to
+infrastructural changes and refactorings in LLVM itself this feature of Rust has
+[been removed](https://github.com/rust-lang/rust/pull/127605) and is no longer
+supported on Nightly at all. As a result there is no longer any possible method
+of writing a function in Rust that returns multiple values at the WebAssembly
+function type level.
+
+In summary this change is expected to not affect any Rust code in the wild
+unless you were using the Nightly feature of `extern "wasm"` in which case
+you'll be forced to drop support for that and use `extern "C"` instead.
+Supporting WebAssembly multi-return functions in Rust is a broader topic than
+this post can cover, but at this time it's an area that's ripe for contribution
+from suitably motivated contributors.
+
+### Aside: ABI Stability and WebAssembly
+
+While on the topics of ABIs and the `multi-value` feature it's perhaps worth
+also going over a bit what ABIs mean for WebAssembly. Currently there's
+effectively only one ABI implemented by LLVM and the Rust compiler meaning that
+at the WebAssembly module level `extern "C"` and `extern "Rust"` are
+more-or-less the same. Despite this though there's no need for this to be true.
+
+The `extern "C"` ABI, what C code uses by default as well, is difficult to
+change because stability is often required across different compiler versions.
+For example WebAssembly code compiled with LLVM 18 might be expected to work
+with code compiled by LLVM 20. This means that changing the ABI is a daunting
+task that requires version fields, explicit markers, etc, to help prevent
+mismatches.
+
+The `extern "Rust"` ABI, however, is entirely within the control of the Rust
```

Comment:
We may want to mention here that `extern "Rust" is considered unstable anyway and this isn't a Wasm-specific issue.
However I was unable to find a specific quote for this in the Rust reference.

Reply:
Rust ABI is perma-unstable AFAIK, there's the crABI initiative but that's not the "Rust" ABI.